src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs

```diff
@@ -492,18 +500,16 @@ private static async Task<T> FinalizeTaskReturningThunk<T>(Continuation continua
 
             continuation.Next = finalContinuation;
 
-            AwaitableProxy awaitableProxy = new AwaitableProxy();
-
             while (true)
             {
-                Continuation headContinuation = UnlinkHeadContinuation(awaitableProxy);
+                Continuation headContinuation = UnlinkHeadContinuation(out var awaitableProxy);
                 await awaitableProxy;
                 Continuation? finalResult = DispatchContinuations(headContinuation, out Exception? ex);
                 if (finalResult != null)
                 {
                     Debug.Assert(finalResult == finalContinuation);
                     if (ex != null)
-                        throw ex;
+                        ExceptionDispatchInfo.Capture(ex).Throw();
```

Comment:
Instead of this I think we can clean things up a bit more by rethrowing (with `throw;`) from `DispatchContinuations`. It should also avoid this level of `ExceptionDispatchInfo`.

Reply:
It also seems a bit redundant to throw - just so that async machine would catch ex and put in the task, but I am not sure there is a way around that unless this method implements async completely by hand. I think that would be too far into diminishing returns for a prototype.

Reply:
rethrowing from within `DispatchContinuations` seems interesting. I will take a look.

Reply:
> It also seems a bit redundant to throw - just so that async machine would catch ex and put in the task, but I am not sure there is a way around that unless this method implements async completely by hand. I think that would be too far into diminishing returns for a prototype.

Yeah, presumably we would end up with some more custom fitted code to handle this... originally I just used a TCS, but it didn't handle flowing all the contexts correctly.
I think we can definitely squeeze some more perf out here in the future -- we currently also seem to pay quite a bit of cost for async machinery that is optimized for cases we never hit here.